src/v/cloud_storage/materialized_resources.cc

```diff
@@ -32,6 +41,104 @@ using namespace std::chrono_literals;
 static constexpr ss::lowres_clock::duration stm_jitter_duration = 10s;
 static constexpr ss::lowres_clock::duration stm_max_idle_time = 60s;
 
+struct device_throughput {
+    size_t read = std::numeric_limits<size_t>::max();
+    size_t write = std::numeric_limits<size_t>::max();
+};
+
+struct throughput_limit {
+    size_t disk_node_throughput_limit = std::numeric_limits<size_t>::max();
+    size_t download_shard_throughput_limit = std::numeric_limits<size_t>::max();
+};
+
+// Compute integer multiplication and division without the integer overflow
+inline size_t muldiv(size_t val, size_t mul, size_t div) {
+    const auto dv = std::lldiv(
+      static_cast<long long>(val), static_cast<long long>(div));
+    return dv.quot * mul + dv.rem * mul / div;
+}
+
+// Get device throughput for the mountpoint
+inline ss::future<device_throughput> get_storage_device_throughput() {
+    try {
+        auto cache_path = config::node().cloud_storage_cache_path().native();
+        auto fs = co_await ss::file_stat(cache_path);
+        auto& queue = ss::engine().get_io_queue(fs.device_id);
+        auto cfg = queue.get_config();
+        auto percent
+          = config::shard_local_cfg().cloud_storage_throughput_limit_percent();
+        if (percent > 0) {
+            // percent == 0 indicates that the throttling is disabled
+            // intentionally
+            co_return device_throughput{
+              .read = muldiv(cfg.read_bytes_rate, 100, percent),
+              .write = muldiv(cfg.write_bytes_rate, 100, percent),
```

Comment:
I'm confused by this: `muldiv (1000, 100, 20) = 5000`. Did you want `muldiv(cfg.rate, percent, 100)`?

Reply:
fixed